File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        return self.dfs(root, float('-inf'))
    
    def dfs(self, current, maxVal):
        if not current: 
            return 0
        
        count = 0;
        
        if current.val >= maxVal:
            count += 1
            maxVal = current.val
            
        count += self.dfs(current.left, maxVal)
        count += self.dfs(current.right, maxVal)
        
        return count

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public static int count = 0;
    public int goodNodes(TreeNode root) {
        
        return dfs(root, Integer.MIN_VALUE);
    }
    
    public int dfs(TreeNode current, int maxVal) {
        if(current == null) {
            return 0;
        }
        count = 0;
        
        if(current.val >= maxVal) {
            count++;
            maxVal = current.val;
        }
        
        count += dfs(current.left, maxVal);
        count += dfs(current.right, maxVal);
            
        return count;
    }
}